File: sdk/node/hpkv-rioc/examples/basic.ts
import { RiocClient, RiocConfig, RiocTlsConfig } from '../src';

async function main() {
  // Parse command line arguments
  const useTls = process.argv.includes('--tls');
  const host = process.argv.includes('--host') ? process.argv[process.argv.indexOf('--host') + 1] : 'localhost';
  const port = process.argv.includes('--port') ? parseInt(process.argv[process.argv.indexOf('--port') + 1]) : 8000;

  // Create client configuration
  const config: RiocConfig = {
    host,
    port,
    timeoutMs: 5000
  };

  // Add TLS configuration if enabled
  if (useTls) {
    config.tls = {
      caPath: process.env.RIOC_CA_PATH || '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/ca.crt',
      certificatePath: process.env.RIOC_CERT_PATH || '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/client.crt',
      keyPath: process.env.RIOC_KEY_PATH || '/workspaces/kernel-high-performance-kv-store/api/rioc/certs/client.key',
      verifyHostname: host,
      verifyPeer: true
    };
  }

  console.log('Client Configuration:');
  console.log(`  Host:      ${config.host}`);
  console.log(`  Port:      ${config.port}`);
  console.log(`  TLS:       ${useTls ? 'enabled' : 'disabled'}`);
  if (useTls) {
    console.log(`  CA Path:   ${config.tls?.caPath}`);
    console.log(`  Cert Path: ${config.tls?.certificatePath}`);
    console.log(`  Key Path:  ${config.tls?.keyPath}`);
  }
  console.log('\nConnecting to server...');
  const startTime = process.hrtime.bigint();

  // Create client
  const client = new RiocClient(config);
  const connectTime = Number(process.hrtime.bigint() - startTime) / 1_000_000;
  console.log(`Connected in ${connectTime.toFixed(2)} ms`);

  try {
    // Test data
    const key = Buffer.from('test_key');
    const initialValue = Buffer.from('initial value');
    const newValue = Buffer.from('updated value');

    // Get initial timestamp
    const timestamp = RiocClient.getTimestamp();
    console.log(`\n1. Inserting record with timestamp ${timestamp}`);
    const insertStart = process.hrtime.bigint();
    client.insert(key, initialValue, timestamp);
    const insertTime = Number(process.hrtime.bigint() - insertStart) / 1_000_000;
    console.log(`Insert successful in ${insertTime.toFixed(2)} ms`);

    // Sleep briefly to ensure timestamp increases
    await new Promise(resolve => setTimeout(resolve, 1));

    console.log('\n2. Getting record');
    const getStart = process.hrtime.bigint();
    const value = client.get(key);
    const getTime = Number(process.hrtime.bigint() - getStart) / 1_000_000;
    if (value === null) {
      console.log(`Key not found (took ${getTime.toFixed(2)} ms)`);
    } else {
      console.log(`Get successful in ${getTime.toFixed(2)} ms, value: ${value.toString()}`);
    }

    // Sleep briefly to ensure timestamp increases
    await new Promise(resolve => setTimeout(resolve, 1));

    // Update
    const updateTimestamp = RiocClient.getTimestamp();
    console.log(`\n3. Updating record with timestamp ${updateTimestamp}`);
    const updateStart = process.hrtime.bigint();
    client.insert(key, newValue, updateTimestamp);
    const updateTime = Number(process.hrtime.bigint() - updateStart) / 1_000_000;
    console.log(`Update successful in ${updateTime.toFixed(2)} ms`);

    // Sleep briefly to ensure timestamp increases
    await new Promise(resolve => setTimeout(resolve, 1));

    console.log('\n4. Getting updated record');
    const getUpdatedStart = process.hrtime.bigint();
    const retrievedValue = client.get(key);
    const getUpdatedTime = Number(process.hrtime.bigint() - getUpdatedStart) / 1_000_000;
    if (retrievedValue === null) {
      console.log(`Key not found (took ${getUpdatedTime.toFixed(2)} ms)`);
    } else {
      console.log(`Get successful in ${getUpdatedTime.toFixed(2)} ms, value: ${retrievedValue.toString()}`);
    }

    // Sleep briefly to ensure timestamp increases
    await new Promise(resolve => setTimeout(resolve, 1));

    // Delete
    const deleteTimestamp = RiocClient.getTimestamp();
    console.log('\n5. Deleting record');
    const deleteStart = process.hrtime.bigint();
    client.delete(key, deleteTimestamp);
    const deleteTime = Number(process.hrtime.bigint() - deleteStart) / 1_000_000;
    console.log(`Delete successful in ${deleteTime.toFixed(2)} ms`);

    // Test get after delete
    console.log('\n6. Getting deleted record');
    const getDeletedStart = process.hrtime.bigint();
    const deletedValue = client.get(key);
    const getDeletedTime = Number(process.hrtime.bigint() - getDeletedStart) / 1_000_000;
    if (deletedValue === null) {
      console.log(`Get after delete correctly returned null in ${getDeletedTime.toFixed(2)} ms`);
    } else {
      console.log(`Unexpected: Get after delete returned value in ${getDeletedTime.toFixed(2)} ms`);
    }
  } finally {
    const disposeStart = process.hrtime.bigint();
    client.dispose();
    const disposeTime = Number(process.hrtime.bigint() - disposeStart) / 1_000_000;
    console.log(`\nAll tests completed successfully (cleanup took ${disposeTime.toFixed(2)} ms)`);
  }
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
}); 

File: sdk/python/hpkv-rioc/examples/basic.py
"""
Basic example of using the HPKV RIOC Python SDK.
"""

import argparse
import os
import time
from datetime import datetime, UTC

from hpkv_rioc import RiocClient, RiocConfig, RiocTlsConfig
from hpkv_rioc.exceptions import RiocError

def get_default_cert_paths():
    """Get default certificate paths from the RIOC certs directory."""
    workspace_root = "/workspaces/kernel-high-performance-kv-store"
    certs_dir = os.path.join(workspace_root, "api/rioc/certs")
    return {
        "client_cert": os.path.join(certs_dir, "client.crt"),
        "client_key": os.path.join(certs_dir, "client.key"),
        "ca_cert": os.path.join(certs_dir, "ca.crt")
    }

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description="RIOC Basic Example")
    cert_paths = get_default_cert_paths()
    
    parser.add_argument("--host", default="localhost",
                      help="Server host")
    parser.add_argument("--port", type=int, default=8000,
                      help="Server port")
    parser.add_argument("--timeout", type=int, default=5000,
                      help="Operation timeout in milliseconds")
    parser.add_argument("--tls", action="store_true",
                      help="Enable TLS")
    parser.add_argument("--client-cert", default=cert_paths["client_cert"],
                      help="Path to client certificate file")
    parser.add_argument("--client-key", default=cert_paths["client_key"],
                      help="Path to client private key file")
    parser.add_argument("--ca-cert", default=cert_paths["ca_cert"],
                      help="Path to CA certificate file")
    
    args = parser.parse_args()

    # Create TLS configuration if enabled
    tls_config = None
    if args.tls:
        # Verify certificate files exist
        for cert_path in [args.client_cert, args.client_key, args.ca_cert]:
            if not os.path.exists(cert_path):
                print(f"Error: Certificate file not found: {cert_path}")
                return

        tls_config = RiocTlsConfig(
            certificate_path=args.client_cert,
            key_path=args.client_key,
            ca_path=args.ca_cert,
            verify_hostname=args.host,
            verify_peer=True
        )

    # Create client configuration
    config = RiocConfig(
        host=args.host,
        port=args.port,
        timeout_ms=args.timeout,
        tls=tls_config
    )

    print(f"Connecting to {args.host}:{args.port}{' with TLS' if args.tls else ''}...")
    if args.tls:
        print(f"Using certificates:")
        print(f"  Client cert: {args.client_cert}")
        print(f"  Client key:  {args.client_key}")
        print(f"  CA cert:     {args.ca_cert}")

    start_time = datetime.now(UTC)

    try:
        # Create client
        client = RiocClient(config)
        connect_time = datetime.now(UTC) - start_time
        print(f"Connected in {connect_time.total_seconds() * 1000:.2f} ms")

        # Test data
        key = "test_key"
        initial_value = "initial value"
        updated_value = "updated value"

        # Get initial timestamp
        timestamp = client.get_timestamp()
        print(f"\n1. Inserting record with timestamp {timestamp}")
        start_time = datetime.now(UTC)
        try:
            client.insert_string(key, initial_value, timestamp)
            insert_time = datetime.now(UTC) - start_time
            print(f"Insert successful in {insert_time.total_seconds() * 1000:.2f} ms")
        except RiocError as e:
            print(f"Insert failed: {e}")
            if e.code == -3:  # I/O error
                print("This usually means the RIOC server is not running or is not accessible.")
                print("Please make sure the server is running and accessible at the specified host and port.")
            return

        # Sleep briefly to ensure timestamp increases
        time.sleep(0.001)

        print("\n2. Getting record")
        start_time = datetime.now(UTC)
        try:
            value = client.get_string(key)
            get_time = datetime.now(UTC) - start_time
            print(f"Get successful in {get_time.total_seconds() * 1000:.2f} ms, value: {value}")
        except RiocError as e:
            get_time = datetime.now(UTC) - start_time
            print(f"Get failed: {e} (took {get_time.total_seconds() * 1000:.2f} ms)")
            return

        # Sleep briefly to ensure timestamp increases
        time.sleep(0.001)

        # Full update
        timestamp = client.get_timestamp()
        print(f"\n3. Updating record with timestamp {timestamp}")
        start_time = datetime.now(UTC)
        try:
            client.insert_string(key, updated_value, timestamp)
            update_time = datetime.now(UTC) - start_time
            print(f"Update successful in {update_time.total_seconds() * 1000:.2f} ms")
        except RiocError as e:
            print(f"Update failed: {e}")
            return

        # Sleep briefly to ensure timestamp increases
        time.sleep(0.001)

        print("\n4. Getting updated record")
        start_time = datetime.now(UTC)
        try:
            value = client.get_string(key)
            get_time = datetime.now(UTC) - start_time
            print(f"Get successful in {get_time.total_seconds() * 1000:.2f} ms, value: {value}")
        except RiocError as e:
            get_time = datetime.now(UTC) - start_time
            print(f"Get failed: {e} (took {get_time.total_seconds() * 1000:.2f} ms)")
            return

        # Sleep briefly to ensure timestamp increases
        time.sleep(0.001)

        # Test delete
        print("\n5. Deleting record")
        timestamp = client.get_timestamp()
        start_time = datetime.now(UTC)
        try:
            client.delete_string(key, timestamp)
            delete_time = datetime.now(UTC) - start_time
            print(f"Delete successful in {delete_time.total_seconds() * 1000:.2f} ms")
        except RiocError as e:
            print(f"Delete failed: {e}")
            return

        # Test get after delete
        print("\n6. Getting deleted record")
        start_time = datetime.now(UTC)
        try:
            value = client.get_string(key)
            get_time = datetime.now(UTC) - start_time
            print(f"Get unexpectedly succeeded in {get_time.total_seconds() * 1000:.2f} ms")
        except RiocError as e:
            get_time = datetime.now(UTC) - start_time
            print(f"Get after delete correctly returned error in {get_time.total_seconds() * 1000:.2f} ms")

        print("\nAll tests completed successfully")

    except RiocError as e:
        print(f"Error: {e}")
        if e.code == -3:  # I/O error
            print("This usually means the RIOC server is not running or is not accessible.")
            print("Please make sure the server is running and accessible at the specified host and port.")
        return

if __name__ == "__main__":
    main() 